TrackingSystem/src/main/java/de/telran/controller/CustomerController.java

```diff
@@ -1,33 +1,79 @@
 package de.telran.controller;
 
+import de.telran.dto.CustomerDTO;
 import de.telran.entity.Customer;
+import de.telran.entity.Shipment;
 import de.telran.service.CustomerService;
+import org.modelmapper.ModelMapper;
 import org.springframework.beans.factory.annotation.Autowired;
-import org.springframework.web.bind.annotation.GetMapping;
-import org.springframework.web.bind.annotation.PostMapping;
-import org.springframework.web.bind.annotation.RequestBody;
-import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.*;
 
 import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
 
 @RestController
 public class CustomerController {
 
     CustomerService customerService;
 
+    ModelMapper modelMapper = new ModelMapper();
+
     @Autowired
     public CustomerController(CustomerService customerService) {
         this.customerService = customerService;
     }
 
     @GetMapping("/api/customers")
-    List<Customer> getAllCustomers() {
-        return customerService.getAllCustomers();
+    List<CustomerDTO> getAllCustomers() {
+        return  customerService.getAllCustomers()
+                .stream()
+                .map(this::convertCustomerToCustomerDTO)
+                .collect(Collectors.toList());
+    }
+
+    @GetMapping("api/customers/{customer_id}/shipments")
+    CustomerDTO getAllShipmentsOfACustomer(@PathVariable("customer_id") Long id) {
+        Customer customer = customerService.getCustomerById(id);
+        return convertCustomerToCustomerDTO(customer);
     }
 
+    @GetMapping("api/customers/{customer_id}/shipments/{shipment_id}/statuses")
+    Optional<Shipment> getCurrentStatusesOfAShipmentOfACustomer(@PathVariable("customer_id") Long customerId,
+                                                                @PathVariable("shipment_id") Long shipmentId) {
+        Customer customer = customerService.getCustomerById(customerId);
+        List<Shipment> shipments = customer.getShipments();
+        return shipments.stream().filter(shipment -> shipment.getShipmentId().equals(shipmentId)).findFirst();
```

Comment:
What you are doing is technically correct: get all shipments of a customer from DB and use Java Stream API to find the one with the specified ID. The problem here is if the customer has hundreds or thousands of shipments, you will have hard performance time filtering in Java. Doing this in SQL is much faster. So, you may need to introduce a method to your service like `getShipmentsByCustomer(Long customerId)`, which in turn calls the Repository with a customer JPQL, which populates a Shipment with its statuses. So, you read a Shipment as we read Customer in previous lesson.